sources/api/migration/migrations/v1.15.0/oci-defaults-max-open-files/src/main.rs

```diff
@@ -0,0 +1,68 @@
+use migration_helpers::common_migrations::NoOpMigration;
+use migration_helpers::{migrate, Migration, MigrationData, Result};
+use serde_json::Value;
+use std::process;
+
+const HARD_RESOURCE_LIMIT_SETTING_NAME: &str =
+    "settings.oci-defaults.resource-limits.max-open-files.hard-limit";
+const SOFT_RESOURCE_LIMIT_SETTING_NAME: &str =
+    "settings.oci-defaults.resource-limits.max-open-files.soft-limit";
+
+/// This migration changes the hard and soft limit for rlimit_nofile to u32 from i64 on downgrade.
+/// There is no need of migration on upgrade as u32 will automatically change to i64
+pub struct ChangeMaxOpenFileResourceLimitType;
+
+fn convert_to_u_32(value: &mut Value) {
+    let s = if value.is_i64() {
+        let v: i64 = serde_json::from_value(value.clone()).unwrap();
+        match v {
+            -1 => u32::MAX,
+            v if v > u32::MAX as i64 => u32::MAX,
+            _ => v as u32,
+        }
+    } else {
+        return;
+    };
+    *value = Value::Number(s.into());
+}
```

Comment:
This doesn't cover all cases. You need to either cover all cases that a `Value` can be, or limit the type to the ones you plan to cover (in this case `i64`). I think we are pretty certain this can only be a few things (mostly `i64` or `u32` since we are storing this in our datastore) so at least you could ensure its a `Number` but then you should also cover the case where could be a floating point. This is a little nitpicky but we should endeavor to use the type system when we can to avoid strange issues.

Reply:
It should suffice to handle `value.is_i64()` and otherwise just leave the value alone, right?
```suggestion
fn convert_to_u_32(value: &mut Value) {
    if !value.is_i64() {
        return;
    }
    let v: i64 = serde_json::from_value(value.clone()).unwrap();
    match v {
        -1 => u32::MAX,
        v if v > u32::MAX as i64 => u32::MAX,
        _ => v as u32,
    }
    *value = Value::Number(s.into());
}
```

Reply:
I agree with @bcressey's suggestion.